File: ios/GildongmuKit/Sources/GildongmuKit/ChatSuggestionsService.swift
import Foundation

/// POST /api/chat/suggestions 요청 본문(서버 zod 스키마 미러).
struct ChatSuggestionsRequestBody: Encodable, Sendable {
    let lastUserMessage: String
    let lastAssistantMessage: String
    let locale: String
    let placeName: String?
}

/// 채팅 답변 뒤 follow-up 질문 제안 조회(spec 2026-08-24 §3.3).
/// `ChatService` 옆 구조체이고 `APIClient` 관례(URLRequest·JSONEncoder)를 따른다.
/// 실패·디코딩 불가·취소·타임아웃 전부 `[]` — 칩은 없어도 되는 보조 컨트롤이라
/// 어떤 실패도 "칩 없음"이지 오류가 아니다.
public struct ChatSuggestionsService: Sendable {
    /// 서버 제한(3개)과 같은 상한. 서버가 더 주더라도 여기서 절단한다.
    public static let maxSuggestions = 3
    /// 본 답변 뒤에 비동기로 붙는 보조 요청이라 예산을 명시한다(APIClient `timeout` 주석 동형).
    public static let timeoutSeconds: TimeInterval = 6

    let session: URLSession

    public init(session: URLSession = .shared) {
        self.session = session
    }

    /// 응답 본문 `{suggestions: string[]}` → 제안 목록. 네트워크 없이 테스트하기 위해 분리한
    /// 순수 함수. `suggestions` 부재·비배열은 `[]`, 비문자열·공백 원소는 건너뛰고
    /// 최대 `maxSuggestions`개로 절단한다.
    public static func parse(_ data: Data) -> [String] {
        guard let object = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let list = object["suggestions"] as? [Any] else {
            return []
        }
        return Array(
            list.lazy
                .compactMap { $0 as? String }
                .map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }
                .filter { !$0.isEmpty }
                .prefix(maxSuggestions)
        )
    }

    public func fetchFollowUps(
        lastUserMessage: String,
        lastAssistantMessage: String,
        locale: String,
        placeName: String?,
        baseURL: URL
    ) async -> [String] {
        var request = URLRequest(url: baseURL.appending(path: "/api/chat/suggestions"))
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.timeoutInterval = Self.timeoutSeconds
        let body = ChatSuggestionsRequestBody(
            lastUserMessage: lastUserMessage,
            lastAssistantMessage: lastAssistantMessage,
            locale: locale,
            placeName: placeName
        )
        guard let encoded = try? JSONEncoder().encode(body) else { return [] }
        request.httpBody = encoded

        guard let (data, response) = try? await session.data(for: request) else { return [] }
        guard !Task.isCancelled else { return [] }
        let status = (response as? HTTPURLResponse)?.statusCode ?? 0
        guard (200..<300).contains(status) else { return [] }
        return Self.parse(data)
    }
}

File: android/kit/src/main/kotlin/space/dodoplanet/gildongmu/kit/ChatSuggestionsService.kt
package space.dodoplanet.gildongmu.kit

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive

/** POST /api/chat/suggestions 요청 본문(서버 zod 스키마 미러). placeName null은 키 생략. */
@Serializable
internal data class ChatSuggestionsRequestBody(
    val lastUserMessage: String,
    val lastAssistantMessage: String,
    val locale: String,
    val placeName: String?,
)

/**
 * 채팅 답변 뒤 follow-up 질문 제안(spec 2026-08-24 §3.3)의 판정 부분. Kit `ChatSuggestionsService.swift` 미러.
 * **POST 전송은 :app(M6)**이 맡는다(D5 경계). 전송 실패·타임아웃·디코딩 불가는 빈 목록(바깥 코루틴 취소는 전파한다) —
 * 칩은 없어도 되는 보조 컨트롤이라 어떤 실패도 "칩 없음"이지 오류가 아니다.
 */
object ChatSuggestionsService {
    const val path = "/api/chat/suggestions"

    /** 서버 제한(3개)과 같은 상한. 서버가 더 주더라도 여기서 절단한다. */
    const val maxSuggestions = 3

    /** 본 답변 뒤에 비동기로 붙는 보조 요청이라 예산을 명시한다(Swift `timeoutSeconds` 6초). */
    const val timeoutMs = 6_000L

    fun encodeBody(lastUserMessage: String, lastAssistantMessage: String, locale: String, placeName: String?): String =
        KitJson.encodeToString(
            ChatSuggestionsRequestBody.serializer(),
            ChatSuggestionsRequestBody(lastUserMessage, lastAssistantMessage, locale, placeName),
        )

    /** 응답 상태·본문 → 제안 목록. 2xx가 아니면 빈 목록. */
    fun followUps(status: Int, body: String): List<String> =
        if (status in 200..299) parse(body) else emptyList()

    /**
     * 응답 본문 `{suggestions: string[]}` → 제안 목록. `suggestions` 부재·비배열은 빈 목록, 비문자열·공백
     * 원소는 건너뛰고 최대 `maxSuggestions`개로 절단한다. 최상위가 객체가 아니거나 JSON이 아니면 빈 목록.
     */
    fun parse(body: String): List<String> {
        val root = runCatching { KitJson.parseToJsonElement(body) }.getOrNull() as? JsonObject ?: return emptyList()
        val list = root["suggestions"] as? JsonArray ?: return emptyList()
        return list.asSequence()
            .mapNotNull { (it as? JsonPrimitive)?.takeIf { p -> p.isString }?.content?.trim() }
            .filter { it.isNotEmpty() }
            .take(maxSuggestions)
            .toList()
    }
}
